staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/maplist.go

```diff
@@ -0,0 +1,191 @@
+	// boolean that indicates whether or not a key could be generated for the provided object.
+	CompositeKeyFor(map[string]interface{}) (interface{}, bool)
+}
+
+// singleKeyStrategy is a cheaper strategy for associative lists that have exactly one key.
+type singleKeyStrategy struct {
+	key     string
+	defawlt interface{} // default is a keyword
+}
+
+// CompositeKeyFor directly returns the value of the single key (or its default value, if absent) to
+// use as a composite key.
+func (ks *singleKeyStrategy) CompositeKeyFor(obj map[string]interface{}) (interface{}, bool) {
+	v, ok := obj[ks.key]
+	if !ok {
+		v = ks.defawlt // substitute default value
+	}
+
+	switch v.(type) {
+	case bool, float64, int64, string:
+		return v, true
+	default:
+		return nil, false // non-scalar
+	}
+}
+
+// hashKeyStrategy computes a hash of all key values.
+type hashKeyStrategy struct {
+	sts    *schema.Structural
+	hasher maphash.Hash
+}
+
+// CompositeKeyFor returns a hash computed from the values (or default values, if absent) of all
+// keys.
+func (ks *hashKeyStrategy) CompositeKeyFor(obj map[string]interface{}) (interface{}, bool) {
+	const keyDelimiter = "\x00" // 0 byte should never appear in the hash input except as delimiter
+
+	ks.hasher.Reset()
+	for _, key := range ks.sts.XListMapKeys {
+		v, ok := obj[key]
+		if !ok {
+			v = ks.sts.Properties[key].Default.Object
+		}
+
+		switch v.(type) {
+		case bool:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%t", v)
+		case float64:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%f", v)
+		case int64:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%d", v)
+		case string:
+			fmt.Fprintf(&ks.hasher, keyDelimiter+"%q", v)
+		default:
+			return nil, false // values must be scalars
+		}
+	}
+	return ks.hasher.Sum64(), true
+}
+
+// emptyMapList is a mapList containing no elements.
+type emptyMapList struct{}
+
+func (emptyMapList) get(interface{}) interface{} {
+	return nil
+}
+
+type mapListImpl struct {
+	sts      *schema.Structural
+	ks       keyStrategy
+	elements map[interface{}][]interface{} // composite key -> bucket
+}
+
+func (a *mapListImpl) get(obj interface{}) interface{} {
+	mobj, ok := obj.(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	key, ok := a.ks.CompositeKeyFor(mobj)
+	if !ok {
+		return nil
+	}
+
+	// Scan bucket to handle key collisions and duplicate key sets:
+	var match interface{}
+	for _, element := range a.elements[key] {
+		all := true
+		for _, key := range a.sts.XListMapKeys {
+			va, ok := element.(map[string]interface{})[key]
+			if !ok {
+				va = a.sts.Properties[key].Default.Object
+			}
+
+			vb, ok := mobj[key]
+			if !ok {
+				vb = a.sts.Properties[key].Default.Object
+			}
+
+			all = all && (va == vb)
+		}
+
+		if !all {
+			continue
+		}
+
+		if match != nil {
+			// Duplicate key set / more than one element matches. This condition should
+			// have generated a validation error elsewhere.
+			return nil
```

Comment:
We can encounter this in oldObj data if it was persisted before `x-kubernetes-list-type: map` was set... I'm not sure what I'd expect to happen in this case. Validation will prevent the new object from being persisted with duplicate keyed list items, but I'm not sure if I'd expect oldSelf to be set to the first matching item in the oldObj or to nil.

Reply:
FWIW, I had switched to keeping 1st encountered value before seeing this comment.